### PR TITLE
refactor: changes after code review

### PR DESCRIPTION
**Pull request by bassosimone**
We're performing a code review of this repository before
writing a new OONI experiment that depends on it.

In this first round of code review, I:

1. fixed all linter warnings in the vpn package;

2. reviewed vpn/bytes.go and vpn/ciphers.go applying minimal
transformations to increase code quality and adding TODO comments
identifying aspects to be discussed with @ainghazal.

See https://github.com/ooni/probe/issues/2083.

### CODE REVIEW

File: vpn/ciphers.go [hunk 1]
```diff
@@ -1,89 +1,152 @@
...
 	}
-	return 0
 }
 
-// decrypt tries to decrypt the ciphertext. ad is optional, and only used in AEAD modes.
-func (a *aesCipher) decrypt(key, iv, ciphertext, ad []byte) ([]byte, error) {
-	k := key[:a.keySizeBytes()] // use stdlib
+// decrypt implements dataCipher.decrypt
+func (a *dataCipherAES) decrypt(key, iv, ciphertext, ad []byte) ([]byte, error) {
+	if len(key) != a.keySizeBytes() {
```
Comment (bassosimone) [inline vpn/ciphers.go:120]:
```suggestion
	if len(key) > a.keySizeBytes() {
```
and we also need to slice the key as before and document what is happening here

File: vpn/ciphers.go [hunk 2]
```diff
@@ -94,93 +157,92 @@ func (a *aesCipher) decrypt(key, iv, ciphertext, ad []byte) ([]byte, error) {
...
-	return nil, nil
 }
 
-// encrypt encrypts the plaintext. ad is optional, and only used in AEAD modes
-func (a *aesCipher) encrypt(key, iv, plaintext, ad []byte) ([]byte, error) {
-	k := key[:a.keySizeBytes()] // get from stdlib
-	i := iv[:a.blockSize()]
+// encrypt implements dataCipher.encrypt
+func (a *dataCipherAES) encrypt(key, iv, plaintext, ad []byte) ([]byte, error) {
+	if len(key) != a.keySizeBytes() {
```
Comment (bassosimone) [inline vpn/ciphers.go:168]: Same as above

File: vpn/ciphers.go [hunk 3]
```diff
@@ -193,14 +255,24 @@ func getHMAC(name string) (func() hash.Hash, bool) {
...
 	return buf[:len(buf)-padding]
 }
 
-// padText does pkcs7 padding of a byte array.
-func padText(buf []byte, bs int) []byte {
+// cipherPadTextPKCS7 does PKCS#7 padding of a byte array.
+func cipherPadTextPKCS7(buf []byte, bs int) []byte {
+	// TODO(bassosimone, ainghazal): what happens if we need to add
+	// no padding? Should we have a special case to handle that in
+	// order to make the code more clear and explicit.
```
[the remark this answers is not in the document]
Comment (bassosimone) [reply vpn/ciphers.go:275]: We should instead have tests

File: vpn/ciphers.go [hunk 4]
```diff
@@ -1,89 +1,152 @@
...
 	}
 
-	var mode cipher.BlockMode
-
 	switch a.mode {
-	case cbcMode:
+	case cipherModeCBC:
 		i := iv[:block.BlockSize()]
-		mode = cipher.NewCBCDecrypter(block, i)
+		mode := cipher.NewCBCDecrypter(block, i)
```
Comment (bassosimone) [inline vpn/ciphers.go:132]: Isn't it strange that the IV is used more than once?! (Maybe I'm just ignorant 😬)

File: vpn/bytes.go [hunk 5]
```diff
@@ -1,25 +1,36 @@
...
+	// TODO(bassosimone,ainghazal): I would expected this function to have
+	// a slightly more descriptive name. What is the name of the encoding
+	// we're actually using here?
+	//
+	// TODO(bassosimone,ainghazal): I am surprised the function that deals
+	// with decoding is not implemented here. I would have expected the
+	// function to encode and the function to decode to be side by side.
+	//
+	// TODO(bassosimone,ainghazal): document this function once we have
+	// figure out what the proper name for it should actually be.
```
[the remark this answers is not in the document]
Comment (kalikaneko) [reply vpn/bytes.go:33]: done

[the remark this answers is not in the document]
Comment (ainghazal) [reply vpn/bytes.go:33]: renamed

File: vpn/ciphers.go (hunk 1 above)
Comment (ainghazal) [reply vpn/ciphers.go:120]: reverted, added comment

File: vpn/ciphers.go (hunk 2 above)
Comment (ainghazal) [reply vpn/ciphers.go:168]: idem

File: vpn/ciphers.go (hunk 4 above)
Comment (ainghazal) [reply vpn/ciphers.go:132]: it's not (see data.go). in AEAD mode, they iv is the packetID + hmac[:8]. In CBC/other, it's just random bytes.

File: vpn/ciphers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ainghazal) [reply vpn/ciphers.go:275]: added a few vectors. we were wrong on our discussion: the spec indicates that you add a whole block size when len modulo bs is zero; in this way it's well behaved and you can use the last byte as indicator of the padding... (but should check for malformed input all the same)

**Comment by ainghazal** [comment]
rebased & merged into main; I think I've addressed all comments so far.

**Comment by bassosimone** [comment]
Closing because code has been merged without using a pull request, thanks!

**Pull request closed without merging**